terraform/projects/app-ecs-services/alertmanager-service.tf

```diff
@@ -106,9 +106,25 @@ resource "aws_ecs_service" "alertmanager_server" {
 
 #### alertmanager
 
+data "pass_password" "pagerduty_service_key" {
+  path = "pagerduty/integration-keys/production"
+}
+
+data "template_file" "alertmanager_config_file" {
+  template = "${file("templates/alertmanager.tpl")}"
+
+  vars {
+    pagerduty_service_key = "${data.pass_password.pagerduty_service_key.password}"
+  }
+}
+
+data "template_file" "alertmanager_dev_config_file" {
+  template = "${file("templates/alertmanager-dev.tpl")}"
+}
+
 resource "aws_s3_bucket_object" "alertmanager" {
-  bucket = "${aws_s3_bucket.config_bucket.id}"
-  key    = "alertmanager/alertmanager.yml"
-  source = "config/alertmanager.yml"
-  etag   = "${md5(file("config/alertmanager.yml"))}"
+  bucket                 = "${aws_s3_bucket.config_bucket.id}"
+  key                    = "alertmanager/alertmanager.yml"
+  content                = "${var.dev_environment == "true" ? data.template_file.alertmanager_dev_config_file.rendered : data.template_file.alertmanager_config_file.rendered}"
+  server_side_encryption = "AES256"
```

Comment:
why has the `etag` gone? can we still ensure that terraform won't re-upload the config file each time?

Reply:
We have done some investigation 🕵️ 

An Etag is generated automatically by AWS when calling `PutObject` (https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html) with encryption so we still get an Etag for our object. We've also checked the Etag of the encrypted object is identical to the Etag of the object uploaded without encryption.

If we add the etag back in as before then we get the following error message:
`Error: aws_s3_bucket_object.nginx-htpasswd: "etag": conflicts with server_side_encryption ("AES256")`.

`Etag` is not a compatible argument if we are defining `kms_key_id` as per the docs (https://www.terraform.io/docs/providers/aws/r/s3_bucket_object.html#etag) but we aren't using `kms` to encrypt and instead using AES256 as our encryption method so we are surprised to see this error message. Maybe is worth us opening an issue with Terraform unless you can think of some areas to further investigate.

We've manually testing running make apply twice and it appears that the files are not reuploaded again given that the "last modified" datetime does not update when you apply the second time.

We also aren't quite sure whether AWS or Terraform is responsible for not reuploading a file if it matches the Etag and haven't been able to find that out either.

Reply:
We've checked and if the file stays the same then terraform does not try and reupload it on top of the existing file.